Sources/Subs-Post.php

```diff
@@ -1310,17 +1312,37 @@ function smtp_mail($mail_to_array, $subject, $message, $headers)
 	if (!server_parse(null, $socket, '220'))
 		return false;
 
-	// Try and determine the servers name, fall back to the mail servers if not found
-	$helo = false;
-	if (function_exists('gethostname') && gethostname() !== false)
-		$helo = gethostname();
-	elseif (function_exists('php_uname'))
-		$helo = php_uname('n');
-	elseif (array_key_exists('SERVER_NAME', $_SERVER) && !empty($_SERVER['SERVER_NAME']))
-		$helo = $_SERVER['SERVER_NAME'];
-
+	// Try to determine the server's fully qualified domain name
+	// Can't rely on $_SERVER['SERVER_NAME'] because it can be spoofed on Apache
 	if (empty($helo))
-		$helo = $modSettings['smtp_host'];
+	{
+		// See if we can get the domain name from the host itself
+		if (function_exists('gethostname'))
+			$helo = gethostname();
+		elseif (function_exists('php_uname'))
+			$helo = php_uname('n');
+
+		// Local domains won't cut it, so look up the canonical one for $boardurl in DNS
+		if (empty($helo) || strpos($helo, '.') === false || substr_compare($helo, '.local', -6) === 0 || (!empty($modSettings['tld_regex']) && !preg_match('/\.' . $modSettings['tld_regex'] . '$/', $helo)))
+		{
+			$helo = parse_url($boardurl, PHP_URL_HOST);
+
+			if (function_exists('gethostbyaddr') && function_exists('dns_get_record'))
+			{
```

Comment:
Shouldn't we test $helo here before going to dns?

Reply:
Meh. I realized that going to DNS was overkill anyway. If the hostname in $boardurl could get a valid result from a DNS lookup, then that hostname was already perfectly good for these purposes. And if it couldn't, it's not like we have any viable options left anyway.